sw/device/lib/dif/dif_i2c.c

```diff
@@ -0,0 +1,661 @@
+                                        I2C_OVRD_TXOVRDEN);
+      break;
+    default:
+      return kDifI2cResultBadArg;
+  }
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_override_drive_pins(const dif_i2c_t *i2c, bool scl,
+                                             bool sda) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t override_val =
+      mmio_region_read32(i2c->base_addr, I2C_OVRD_REG_OFFSET);
+  override_val =
+      bitfield_set_field32(override_val, (bitfield_field32_t){
+                                             .mask = 1,
+                                             .index = I2C_OVRD_SCLVAL,
+                                             .value = scl,
+                                         });
+  override_val =
+      bitfield_set_field32(override_val, (bitfield_field32_t){
+                                             .mask = 1,
+                                             .index = I2C_OVRD_SDAVAL,
+                                             .value = sda,
+                                         });
+  mmio_region_write32(i2c->base_addr, I2C_OVRD_REG_OFFSET, override_val);
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_override_sample_pins(const dif_i2c_t *i2c,
+                                              uint16_t *scl_samples,
+                                              uint16_t *sda_samples) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t samples = mmio_region_read32(i2c->base_addr, I2C_VAL_REG_OFFSET);
+  samples >>= I2C_VAL_SCL_RX_OFFSET;
+  if (scl_samples != NULL) {
+    uint32_t val = samples >> I2C_VAL_SCL_RX_OFFSET;
+    *scl_samples = val & I2C_VAL_SCL_RX_MASK;
+  }
+
+  if (sda_samples != NULL) {
+    uint32_t val = samples >> I2C_VAL_SDA_RX_OFFSET;
+    *sda_samples = val & I2C_VAL_SDA_RX_MASK;
+  }
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_get_fifo_levels(const dif_i2c_t *i2c,
+                                         uint8_t *fmt_fifo_level,
+                                         uint8_t *rx_fifo_level) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t values =
+      mmio_region_read32(i2c->base_addr, I2C_FIFO_STATUS_REG_OFFSET);
+  if (fmt_fifo_level != NULL) {
+    uint32_t val = values >> I2C_FIFO_STATUS_FMTLVL_OFFSET;
+    *fmt_fifo_level = val & I2C_FIFO_STATUS_FMTLVL_MASK;
+  }
+  if (rx_fifo_level != NULL) {
+    uint32_t val = values >> I2C_FIFO_STATUS_RXLVL_OFFSET;
+    *rx_fifo_level = val & I2C_FIFO_STATUS_RXLVL_MASK;
+  }
+
+  return kDifI2cResultOk;
+}
+
+dif_i2c_result_t dif_i2c_read_rx_fifo(const dif_i2c_t *i2c, uint8_t *byte) {
+  if (i2c == NULL) {
+    return kDifI2cResultBadArg;
+  }
+
+  uint32_t rdata = mmio_region_read32(i2c->base_addr, I2C_RDATA_REG_OFFSET);
+  if (byte != NULL) {
+    *byte = (rdata >> I2C_RDATA_RDATA_OFFSET) & I2C_RDATA_RDATA_MASK;
+  }
+
+  return kDifI2cResultOk;
+}
+
```

Comment:
Can these be converted to use the bitfield accessors instead of shifting and masking?

Reply:
The rx one, yeah, but the fifo levels one would require doing two reads. I might add a get_field32 variant to bitfield.h though.